### ISSUE TITLE
MP4 not playing within Webkit

### ISSUE THREAD

**Issue by gil0109** [Igalia/meta-webkit#18]
We create an MP4 using powerpoint presentation to display on our digital signage.  The MP4 is approx. 700mb.  Webkit launches the webpage but does not play the .mp4.

When I point cog to the mp4, it says loading, then error in the player.  Is there a way to see what the error is?

**Comment by clopez** [comment]
Can you play the .mp4 file with the command ```gst-play-1.0``` ?

if not then likely you are missing the decoders for it. If you want software decoders simply install in the image the package gstreamer1.0-libav. If you want hardware-based decoders then it depends on your hardware

**Comment by gil0109** [comment]
Yes, it plays great on gst-play-1.0.

**Comment by clopez** [comment]
That's strange. 

Can you play the mp4 video on this [test example page](https://people.igalia.com/clopez/wkbug/video/simplevideo.html) with cog? If not, double-check that you are passing the video option in the list of values for the wpewebkit PACKAGECONFIG

**Comment by gil0109** [comment]
Yes that one worked. 

Although I can't give the HTML page that opens the MP4, I can give you the url to the mp4.

https://sbc-qsystem-dev.pathfinder.gov.bc.ca/sbc.mp4

It works on a raspberry pi with Raspbian Stretch and chromium no problem.  

Any chance, we can see the error logs?

**Comment by clopez** [comment]
Which wpebackend are you using?

**Comment by gil0109** [comment]
This is my local.conf setup:

MACHINE = "raspberrypi3"
GPU_MEM_256 = "128"
GPU_MEM_512 = "196"
GPU_MEM_1024 = "396"
EXTRA_IMAGE_FEATURES = "debug-tweaks"
IMAGE_FEATURES_append = " ssh-server-dropbear hwcodecs"
PREFERRED_PROVIDER_virtual/wpebackend = "wpebackend-rdk"
PACKAGECONFIG_pn-wpebackend-rdk = "rpi"
IMAGE_INSTALL_append = " wpewebkit cog"
DISTRO_FEATURES_append = " opengl wayland"
IMAGE_INSTALL_append = " gstreamer1.0-omx"
LICENSE_FLAGS_WHITELIST_append = " commercial"

**Comment by clopez** [comment]
Some comments about the videos themselves ... that is likely unrelated with the issue, but in any case i think is worth commenting:

That videos can be optimized by processing them with the command ```qt-faststart``` to include the header at the start of the file, that will make streaming easier/faster for any player.

You can check the current video is not "streamable"
```
$ wget https://sbc-qsystem-dev.pathfinder.gov.bc.ca/sbc_small.mp4

$ mediainfo -f sbc.mp4 |grep -i streama
IsStreamable                             : No
```
You can fix that, by processing them with qt-faststart
```
# After running via qt-faststart
$ qt-faststart sbc.mp4 sbc_fast.mp4
ftyp          0 24
uuid         24 40
mdat         64 644076521
moov  644076585 170625
 patching stco atom...
 writing ftyp atom...
 writing moov atom...
 copying rest of file...

$ mediainfo -f sbc_fast.mp4 |grep -i streama
IsStreamable                             : Yes
```

**Comment by clopez** [comment]
I have tried to play the videos with cog and wpewebkit and it worked fine for me, but I'm testing on the desktop. On the RPi it may be different as on the RPi you are using the hardware-based gstreamer-omx decoders.



> Any chance, we can see the error logs?

To get the gstreamer error logs you can export this environment variables before starting cog
```
export GST_DEBUG="*:3"
```
To get even more logging, [raise the number as indicated here](https://gstreamer.freedesktop.org/documentation/tutorials/basic/debugging-tools.html#the-debug-log)

**Comment by gil0109** [comment]
do you think its because of the size of the rootfs?  Is there an easy way to increase the size?

**Comment by clopez** [comment]
> do you think its because of the size of the rootfs? Is there an easy way to increase the size?

Now that you mention it, if there isn't space enough to temporally store the video file it can cause issues.

Try setting on local.conf something like this:

```
# Add 5GB of extra free space to the image
IMAGE_ROOTFS_EXTRA_SPACE = "5242880"
```
Ref: https://www.yoctoproject.org/docs/current/mega-manual/mega-manual.html#var-IMAGE_ROOTFS_EXTRA_SPACE

**Comment by gil0109** [comment]
Converted file and tried again with image including extra rootfs space.   Still doesn't work.

https://sbc-qsystem-dev.pathfinder.gov.bc.ca/sbc_new.mp4

Any other ideas? 

Here is the debugging output.

![image](https://user-images.githubusercontent.com/19917617/41950515-3889a334-797b-11e8-8d3f-cefa4c53418c.png)

**Comment by gil0109** [comment]
Your instincts were right.  That is not the issue.  Problem still exists.
![image](https://user-images.githubusercontent.com/19917617/41951697-101a1a86-7981-11e8-9f15-b301a4a79a62.png)

**Comment by clopez** [comment]
Maybe some more useful error gets logged with ```export GST_DEBUG="*:5" ```? (or 4 instead of 5 if that is too verbose) ?

**Comment by gil0109** [comment]
Is there an easy way to pipe the log output to a file?

**Comment by gil0109** [comment]
I figured how to pipe to file.  I hope this helps...
[karim.txt](https://github.com/Igalia/meta-webkit/files/2139895/karim.txt)

**Comment by gil0109** [comment]
Also, tried with weston (Mesa VC drivers) and got this error:

![image](https://user-images.githubusercontent.com/19917617/41956569-eec4d534-7998-11e8-840b-8e5e635cc34f.png)

**Comment by gil0109** [comment]
I followed the instructions from the wiki which uses rocko

**Comment by philn** [comment]
Can you try with `gst-play-1.0 --videosink glimagesink` too?
If this is a RPi, have you made sure to allocate enough men for the GPU at boot? https://www.raspberrypi.org/documentation/configuration/config-txt/memory.md
Do you use wpebackend-rdk/rpi-dispmanx or wpe-backend-fdo with VC4? I doubt the latter works yet for video playback.

**Comment by philn** [comment]
If there are buffering issues this is likely a WebKit issue..
Can you try this? `GST_DEBUG_FILE=gst.log GST_DEBUG_NO_COLOR=1 GST_DEBUG=webkit*:6 cog ...` and attach gst.log here?

**Comment by philn** [comment]
Another thing to try, disable this code: https://github.com/WebKit/webkit/blob/master/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp#L162-L166

Then another element will be used to read data from the http server, likely souphttpsrc.

**Comment by philn** [comment]
https://github.com/WebKit/webkit/blob/master/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp#L164 needs to be kept though.

**Comment by gil0109** [comment]
GST_DEBUG_FILE=gst.log GST_DEBUG_NO_COLOR=1 GST_DEBUG=webkit*:6 cog results:
[gst.log](https://github.com/Igalia/meta-webkit/files/2142973/gst.log)

For the testing of the changes to the code, I will need to do them in a few hours.

**Comment by gil0109** [comment]
Created local video.html file.  Launched with the following command: cog file:video.html.  And it worked fine.

**Comment by clopez** [comment]
AFAIK You can't modify the source files directly, you have to create a patch and add it to the recipe in the SRC_URI list

**Comment by gil0109** [comment]
Any chance you can walk me through the steps?

guessing I start with: devtool ?

**Comment by gil0109** [comment]
I think I figured it out using this: https://wiki.yoctoproject.org/wiki/TipsAndTricks/Patching_the_source_for_a_recipe

I will use this recipe:  devtool modify wpewebkit

**Comment by gil0109** [comment]
@clopez 

I made the changes and rebuilding with the following changes, please confirm:

#    GRefPtr<GstElementFactory> srcFactory = adoptGRef(gst_element_factory_find("webkitwebsrc"));
#    if (!srcFactory) {
        GST_DEBUG_CATEGORY_INIT(webkit_media_player_debug, "webkitmediaplayer", 0, "WebKit media playe$
#        gst_element_register(0, "webkitwebsrc", GST_RANK_PRIMARY + 100, WEBKIT_TYPE_WEB_SRC);
#    }

    return true;

**Comment by gil0109** [comment]
Would you like me to build with sumo or master instead?

![image](https://user-images.githubusercontent.com/19917617/42011946-c6403f4c-7a4a-11e8-8a7d-a7f90ec7c93e.png)

**Comment by philn** [comment]
Comments in C/C++ are either 
`// this is a one-line comment`

or `/* multiple
lines */`

**Comment by gil0109** [comment]
@philn @clopez  thanks for the tips.  

Sadly, the removal of those lines did not resolve the issue.  The mp4 player controls that shows up, and instead of staying loading (for 20sec) then error, it goes straight to error.

I appreciate all your help, any ideas what to do next?

**Comment by gil0109** [comment]
I also built without GPU Acceleration and that didn't seem to make any difference.

Removing these lines from local.conf:

DISTRO_FEATURES_append = " opengl wayland"
IMAGE_INSTALL_append = " gstreamer1.0-omx"
LICENSE_FLAGS_WHITELIST_append = " commercial"

**Comment by philn** [comment]
If an error happens it might appear in the GStreamer logs. Please attach generate a log again.

**Comment by gil0109** [comment]
Do you want it with the image with the "fix" and to confirm, I am getting the logs by entering the following command prior to going to the mp4 file:

GST_DEBUG_FILE=gst.log GST_DEBUG_NO_COLOR=1 GST_DEBUG=webkit*:6 cog

**Comment by gil0109** [comment]
@philn Here is the log attached
[gst.log](https://github.com/Igalia/meta-webkit/files/2146348/gst.log)

**Comment by philn** [comment]
Ah, sorry, forgot to mention this line should also be commented out: https://github.com/WebKit/webkit/blob/master/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L237

**Comment by gil0109** [comment]
Updated and rebuilt.  This time, the video control didn't show up.  The logs are empty but still no video playing.

**Comment by gil0109** [comment]
I went to the page (restricted from the outside) to get details.  Attached is the log for your review @philn 
[gst.log](https://github.com/Igalia/meta-webkit/files/2146930/gst.log)

**Comment by philn** [comment]
Logs are empty, really?

**Comment by gil0109** [comment]
I know.. it was strange.  But there is no video control display either (with the play button.  I will let it sit for 5 min and check the logs again.

**Comment by philn** [comment]
Error 12: Your GStreamer installation is missing a plug-in. (url=https://sbc-qsystem-dev.pathfinder.gov.bc.ca/qsystem/qsmartboard/resource/invite.mp3)

**Comment by philn** [comment]
If you try to play an audio file, use `<audio controls ...>` for video... use `<video controls ...>`.

**Comment by gil0109** [comment]
The sound playing is from another iframe.  I am not concerned about playing that.  When I do, I will fix it.

The video is in a separate iframe (url: vid2.html).  Note, I pointed to the mp4 directly and it crashed on the screen but still no logs.

**Comment by philn** [comment]
Same error for the mp4... Please make sure your GStreamer installation is ok... Use `gst-play-1.0` to check playback outside of webkit.

**Comment by gil0109** [comment]
You are correct, I will check my layer.conf and rebuild later today and report back.  Thanks again for your patience.

**Comment by gil0109** [comment]
That worked!!  Do I create my own patch, or will you do a pull request to fix it moving forward?

**Comment by philn** [comment]
No it was just a test to verify this is a WebKit bug. See https://bugs.webkit.org/show_bug.cgi?id=187171

**Comment by gil0109** [comment]
Great, thank you for your help on this.  It's been a pleasure working with you.

**Comment by gil0109** [comment]
@philn A couple of followup questions, if you don't mind:
1. Will this issue be closed one the bug fix has been applied to webkit?
2. What is the temp folder that cog uses?  Is there a way to move it as I am making the rootfs read-only

**Comment by philn** [comment]
Yes this issue will be closed once the wpe recipe contains the webkit fix. About the cog temp folder, I'm not sure, I opened an issue: https://github.com/Igalia/cog/issues/44

**Comment by gil0109** [comment]
@philn I think this issue is resolved.  I rebuild from scratch without modifying lines above and it seems to be working now.  Not sure if they fixed it with another pull request.

**Comment by gil0109** [comment]
@philn I rebuilt the image again today and all seems to be working without the changes mentioned above.  Not sure if they fixed it on their end but I am good now.  So I am not sure if we want to close the issue with webkit and this one.  I will leave that up to you.

Up next: I am now trying to run with these variables set:

export XDG_CACHE_HOME="/tmp"
export XDG_DATA_HOME="/tmp"

Which is causing the image to be "corrupt" after about 15sec of playback.  I am thinking it's an issue with my tmpfs size.  Nothing is easy ;)

**Comment by gil0109** [comment]
@philn Thank you for all your help :)

**Comment by Ankur498** [comment]
Please help me on this quite a long time i am struggling with this 
I am getting below error 

root@raspberrypi-rdk-mc:~# gst-launch-1.0 playbin uri=http://35.198.234.156:9003/vr/
Setting pipeline to PAUSED ...
Pipeline is PREROLLING ...
Missing element: text/html decoder
WARNING: from element /GstPlayBin:playbin0/GstURIDecodeBin:uridecodebin0: No decoder available for type 'text/html'.
Additional debug info:
../../../gst-plugins-base-1.10.4/gst/playback/gsturidecodebin.c(938): unknown_type_cb (): /GstPlayBin:playbin0/GstURIDecodeBin:uridecodebin0
ERROR: from element /GstPlayBin:playbin0/GstURIDecodeBin:uridecodebin0: Your GStreamer installation is missing a plug-in.
Additional debug info:
../../../gst-plugins-base-1.10.4/gst/playback/gsturidecodebin.c(1005): no_more_pads_full (): /GstPlayBin:playbin0/GstURIDecodeBin:uridecodebin0:
no suitable plugins found:
../../../gst-plugins-base-1.10.4/gst/playback/gstdecodebin2.c(4592): gst_decode_bin_expose (): /GstPlayBin:playbin0/GstURIDecodeBin:uridecodebin0/GstDecodeBin:decodebin0:
no suitable plugins found:
Missing decoder: text/html (text/html)

ERROR: pipeline doesn't want to preroll.
Setting pipeline to NULL ...
Freeing pipeline ...
root@raspberrypi-rdk-mc:~#



I tried wpeLauncher also no video is getting loaded .

I am little naive on RDK , please let me know how to enable text/html decoder
